[dhalion] null exception for previousCheckpoint

We saw the below exception in the HealthMgr log. The exception was from Dhalion library. The exception caused the HealthMgr process to quit.

```
[2018-06-20 20:44:11 +0000] [STDERR] stderr: Exception in thread "main"   
[2018-06-20 20:44:11 +0000] [STDERR] stderr: java.util.concurrent.ExecutionException: java.lang.NullPointerException  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at org.apache.heron.healthmgr.HealthManager.main(HealthManager.java:241)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: Caused by: java.lang.NullPointerException  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.time.Instant.compareTo(Instant.java:1255)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.time.Instant.isBefore(Instant.java:1285)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at com.microsoft.dhalion.policy.PoliciesExecutor.lambda$null$0(PoliciesExecutor.java:83)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at com.microsoft.dhalion.policy.PoliciesExecutor.lambda$start$2(PoliciesExecutor.java:84)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)  
[2018-06-20 20:44:11 +0000] [STDERR] stderr: 	at java.lang.Thread.run(Thread.java:748)  
```

I doubt the Null was from https://github.com/Microsoft/Dhalion/blob/0.2.1/src/main/java/com/microsoft/dhalion/policy/PoliciesExecutor.java#L74?

Btw, we have several jobs and saw this issue in only one job.

Hi @avflor 

The healthmgr failed and quited in the first round policy execution, then restarted. There were no last 5 policy execution.

The issue is `previousCheckpoint ` is not initialized (null) in the first round policy execution. https://github.com/Microsoft/Dhalion/blob/0.2.1/src/main/java/com/microsoft/dhalion/policy/PoliciesExecutor.java#L137


```
[2018-06-27 23:15:11 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Logging setup done.  
[2018-06-27 23:15:11 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Static Heron config loaded successfully   
[2018-06-27 23:15:11 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Initializing health manager  
[2018-06-27 23:15:13 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Starting Curator client connecting to: xxxxxxx_zk
[2018-06-27 23:15:13 +0000] [INFO] org.apache.curator.framework.imps.CuratorFrameworkImpl: Starting  
[2018-06-27 23:15:13 +0000] [INFO] org.apache.curator.framework.state.ConnectionStateManager: State change: CONNECTED  
[2018-06-27 23:15:13 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Directory tree initialized.  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Starting Curator client connecting to: xxxxxxx_zk 
[2018-06-27 23:15:14 +0000] [INFO] org.apache.curator.framework.imps.CuratorFrameworkImpl: Starting  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.curator.framework.state.ConnectionStateManager: State change: CONNECTED  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Directory tree initialized.  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.HealthPolicyConfig: Health Policy Configuration:{health.policy.class=org.apache.heron.healthmgr.policy.AutoRestartBackpressureContainerPolicy, health.policy.interval.ms=120000, BackPressureDetector.noiseFilterMillis=20}  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Initializing auto-restart-backpressure-container with class org.apache.heron.healthmgr.policy.AutoRestartBackpressureContainerPolicy  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.sensors.MetricsCacheMetricsProvider: Metrics will be provided by MetricsCache at http://yyyyyyyyyy/stats  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Starting Health Manager  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Starting Health Manager metric posting thread  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.common.network.HeronClient: Connecting to endpoint: /127.0.0.1:31671  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.HealthManagerMetrics: Connected to Metrics Manager. Ready to send register request  
[2018-06-27 23:15:14 +0000] [INFO] com.microsoft.dhalion.policy.PoliciesExecutor: Executing Policy: AutoRestartBackpressureContainerPolicy, checkpoint: 2018-06-27T23:15:14.462Z  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.common.TopologyProvider: Fetching topology from state manager: zzzzzzzzzz  
[2018-06-27 23:15:14 +0000] [INFO] org.apache.heron.healthmgr.HealthManagerMetrics: We registered ourselves to the Metrics Manager  
[2018-06-27 23:15:15 +0000] [INFO] org.apache.heron.common.network.SocketChannelHelper: Forcing to flush data to socket with best effort.  
[2018-06-27 23:15:15 +0000] [INFO] org.apache.heron.common.network.HeronClient: To stop the HeronClient.  
[2018-06-27 23:15:15 +0000] [INFO] org.apache.heron.healthmgr.HealthManagerMetrics: SimpleMetricsManagerClient exits  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: Exception in thread "main"   
[2018-06-27 23:15:15 +0000] [STDERR] stderr: java.util.concurrent.ExecutionException: java.lang.NullPointerException  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at org.apache.heron.healthmgr.HealthManager.main(HealthManager.java:241)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: Caused by: java.lang.NullPointerException  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.time.Instant.compareTo(Instant.java:1255)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.time.Instant.isBefore(Instant.java:1285)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at com.microsoft.dhalion.policy.PoliciesExecutor.lambda$null$0(PoliciesExecutor.java:83)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at com.microsoft.dhalion.policy.PoliciesExecutor.lambda$start$2(PoliciesExecutor.java:84)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)  
[2018-06-27 23:15:15 +0000] [STDERR] stderr: 	at java.lang.Thread.run(Thread.java:748)  
[2018-06-27 23:15:26 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Logging setup done.  
[2018-06-27 23:15:26 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Static Heron config loaded successfully   
[2018-06-27 23:15:26 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Initializing health manager  
[2018-06-27 23:15:27 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Starting Curator client connecting to: xxxxxxx_zk
[2018-06-27 23:15:27 +0000] [INFO] org.apache.curator.framework.imps.CuratorFrameworkImpl: Starting  
[2018-06-27 23:15:27 +0000] [INFO] org.apache.curator.framework.state.ConnectionStateManager: State change: CONNECTED  
[2018-06-27 23:15:27 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Directory tree initialized.  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Starting Curator client connecting to: xxxxxxx_zk
[2018-06-27 23:15:28 +0000] [INFO] org.apache.curator.framework.imps.CuratorFrameworkImpl: Starting  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.curator.framework.state.ConnectionStateManager: State change: CONNECTED  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.heron.statemgr.zookeeper.curator.CuratorStateManager: Directory tree initialized.  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.heron.healthmgr.HealthPolicyConfig: Health Policy Configuration:{health.policy.class=org.apache.heron.healthmgr.policy.AutoRestartBackpressureContainerPolicy, health.policy.interval.ms=120000, BackPressureDetector.noiseFilterMillis=20}  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Initializing auto-restart-backpressure-container with class org.apache.heron.healthmgr.policy.AutoRestartBackpressureContainerPolicy  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.heron.healthmgr.sensors.MetricsCacheMetricsProvider: Metrics will be provided by MetricsCache at http://yyyyyyyyyy/stats  
[2018-06-27 23:15:28 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Starting Health Manager  
[2018-06-27 23:15:29 +0000] [INFO] org.apache.heron.healthmgr.HealthManager: Starting Health Manager metric posting thread  
[2018-06-27 23:15:29 +0000] [INFO] org.apache.heron.common.network.HeronClient: Connecting to endpoint: /127.0.0.1:31671  
[2018-06-27 23:15:29 +0000] [INFO] org.apache.heron.healthmgr.HealthManagerMetrics: Connected to Metrics Manager. Ready to send register request  
[2018-06-27 23:15:29 +0000] [INFO] com.microsoft.dhalion.policy.PoliciesExecutor: Executing Policy: AutoRestartBackpressureContainerPolicy, checkpoint: 2018-06-27T23:15:29.036Z  
[2018-06-27 23:15:29 +0000] [INFO] org.apache.heron.healthmgr.common.TopologyProvider: Fetching topology from state manager: zzzzzzzzzz  
[2018-06-27 23:15:29 +0000] [INFO] org.apache.heron.healthmgr.HealthManagerMetrics: We registered ourselves to the Metrics Manager  
[2018-06-27 23:15:30 +0000] [INFO] org.apache.heron.common.network.SocketChannelHelper: Forcing to flush data to socket with best effort.  
[2018-06-27 23:15:30 +0000] [INFO] org.apache.heron.common.network.HeronClient: To stop the HeronClient.  
[2018-06-27 23:15:30 +0000] [INFO] org.apache.heron.healthmgr.HealthManagerMetrics: SimpleMetricsManagerClient exits  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: Exception in thread "main"   
[2018-06-27 23:15:30 +0000] [STDERR] stderr: java.util.concurrent.ExecutionException: java.lang.NullPointerException  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at org.apache.heron.healthmgr.HealthManager.main(HealthManager.java:241)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: Caused by: java.lang.NullPointerException  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.time.Instant.compareTo(Instant.java:1255)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.time.Instant.isBefore(Instant.java:1285)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at com.microsoft.dhalion.policy.PoliciesExecutor.lambda$null$0(PoliciesExecutor.java:83)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at com.microsoft.dhalion.policy.PoliciesExecutor.lambda$start$2(PoliciesExecutor.java:84)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)  
[2018-06-27 23:15:30 +0000] [STDERR] stderr: 	at java.lang.Thread.run(Thread.java:748)  

```

--- a/WORKSPACE
+++ b/WORKSPACE
@@ -460,7 +460,7 @@ maven_jar(
 
 maven_jar(
   name = "com_microsoft_dhalion",
-  artifact = "com.microsoft.dhalion:dhalion:0.2.1",
+  artifact = "com.microsoft.dhalion:dhalion:0.2.3",
 )
 
 maven_jar(

--- a/heron/healthmgr/src/java/org/apache/heron/healthmgr/policy/DynamicResourceAllocationPolicy.java
+++ b/heron/healthmgr/src/java/org/apache/heron/healthmgr/policy/DynamicResourceAllocationPolicy.java
@@ -21,11 +21,13 @@ package org.apache.heron.healthmgr.policy;
 
 import java.time.Duration;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.logging.Logger;
 
 import javax.inject.Inject;
 
 import com.microsoft.dhalion.api.IResolver;
+import com.microsoft.dhalion.core.Action;
 import com.microsoft.dhalion.core.Diagnosis;
 import com.microsoft.dhalion.core.DiagnosisTable;
 import com.microsoft.dhalion.events.EventHandler;
@@ -92,7 +94,7 @@ public class DynamicResourceAllocationPolicy extends HealthPolicyImpl
   }
 
   @Override
-  public IResolver selectResolver(Collection<Diagnosis> diagnosis) {
+  public Collection<Action> executeResolvers(Collection<Diagnosis> diagnosis) {
     DiagnosisTable diagnosisTable = DiagnosisTable.of(diagnosis);
 
     if (diagnosisTable.type(DIAGNOSIS_DATA_SKEW.text()).size() > 0) {
@@ -100,10 +102,10 @@ public class DynamicResourceAllocationPolicy extends HealthPolicyImpl
     } else if (diagnosisTable.type(DIAGNOSIS_SLOW_INSTANCE.text()).size() > 0) {
       LOG.warning("Slow Instance diagnoses. This diagnosis does not have any resolver.");
     } else if (diagnosisTable.type(DIAGNOSIS_UNDER_PROVISIONING.text()).size() > 0) {
-      return scaleUpResolver;
+      return scaleUpResolver.resolve(diagnosis);
     }
 
-    return null;
+    return Collections.EMPTY_LIST;
   }
 
   @Override

--- a/heron/healthmgr/src/java/org/apache/heron/healthmgr/policy/DynamicResourceAllocationPolicy.java
+++ b/heron/healthmgr/src/java/org/apache/heron/healthmgr/policy/DynamicResourceAllocationPolicy.java
@@ -26,7 +26,6 @@ import java.util.logging.Logger;
 
 import javax.inject.Inject;
 
-import com.microsoft.dhalion.api.IResolver;
 import com.microsoft.dhalion.core.Action;
 import com.microsoft.dhalion.core.Diagnosis;
 import com.microsoft.dhalion.core.DiagnosisTable;

--- a/heron/healthmgr/src/java/org/apache/heron/healthmgr/policy/ToggleablePolicy.java
+++ b/heron/healthmgr/src/java/org/apache/heron/healthmgr/policy/ToggleablePolicy.java
@@ -21,6 +21,7 @@ package org.apache.heron.healthmgr.policy;
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.logging.Logger;
 
 import javax.inject.Inject;
@@ -97,7 +98,7 @@ public class ToggleablePolicy extends HealthPolicyImpl {
     if (policyMode.equals(PolicyMode.activated)) {
       return super.executeSensors();
     } else {
-      return new ArrayList<Measurement>();
+      return Collections.EMPTY_LIST;
     }
   }
 
@@ -106,7 +107,7 @@ public class ToggleablePolicy extends HealthPolicyImpl {
     if (policyMode.equals(PolicyMode.activated)) {
       return super.executeDetectors(measurements);
     } else {
-      return new ArrayList<Symptom>();
+      return Collections.EMPTY_LIST;
     }
   }
 
@@ -115,7 +116,7 @@ public class ToggleablePolicy extends HealthPolicyImpl {
     if (policyMode.equals(PolicyMode.activated)) {
       return super.executeDiagnosers(symptoms);
     } else {
-      return new ArrayList<Diagnosis>();
+      return Collections.EMPTY_LIST;
     }
   }
 

--- a/heron/healthmgr/src/java/org/apache/heron/healthmgr/resolvers/RestartContainerResolver.java
+++ b/heron/healthmgr/src/java/org/apache/heron/healthmgr/resolvers/RestartContainerResolver.java
@@ -20,10 +20,9 @@
 package org.apache.heron.healthmgr.resolvers;
 
 import java.time.Instant;
-import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashSet;
-import java.util.List;
 import java.util.logging.Logger;
 
 import javax.inject.Inject;
@@ -74,7 +73,6 @@ public class RestartContainerResolver implements IResolver {
   public Collection<Action> resolve(Collection<Diagnosis> diagnosis) {
     publishingMetrics.executeResolver(RESTART_CONTAINER_RESOLVER);
 
-    List<Action> actions = new ArrayList<>();
 
     // find all back pressure measurements reported in this execution cycle
     Instant current = context.checkpoint();
@@ -85,7 +83,7 @@ public class RestartContainerResolver implements IResolver {
 
     if (bpSymptoms.size() == 0) {
       LOG.fine("No back-pressure measurements found, ending as there's nothing to fix");
-      return actions;
+      return Collections.EMPTY_LIST;
     }
 
     Collection<String> allBpInstances = new HashSet<>();
@@ -116,7 +114,6 @@ public class RestartContainerResolver implements IResolver {
     LOG.info("Broadcasting container restart event");
     ContainerRestart action = new ContainerRestart(current, stmgrIds);
     eventManager.onEvent(action);
-    actions.add(action);
-    return actions;
+    return Collections.singletonList(action);
   }
 }

